feat: add `ndarray/base/mskfilter` (#2604)
stdlib-js/stdlib: lib/node_modules/@stdlib/ndarray/base/mskfilter/README.md

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# mskfilter

> Apply a mask to a provided input ndarray and assign unmasked values to elements in a provided one-dimensional output ndarray.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var mskfilter = require( '@stdlib/ndarray/base/mskfilter' );
```

#### mskfilter( arrays )

> Apply a mask to a provided input ndarray and assign unmasked values to elements in a provided one-dimensional output ndarray.

Reviewer: ```suggestion
Applies a mask to a provided input ndarray and assigns unmasked values to elements in a provided one-dimensional output ndarray.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var Uint8Array = require( '@stdlib/array/uint8' );

// Define the shape of the input and mask arrays:
var shape = [ 3, 1, 2 ];

// Define the array strides:
var sx = [ 2, 2, 1 ];
var sm = [ 2, 2, 1 ];
var sy = [ 1 ];

// Define the index offsets:
var ox = 1;
var om = 0;

// Create the input, mask and output ndarray-like objects:
var x = {
    'dtype': 'float64',

>     'data': new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0 ] ),

Reviewer: Why does this array have 12 elements?

Reviewer: May be better to have only 7 (ie, the min number necessary to accommodate shape, strides, and offset).

    'shape': shape,
    'strides': sx,
    'offset': ox,
    'order': 'row-major'
};
var mask = {
    'dtype': 'uint8',
    'data': new Uint8Array( [ 1, 0, 1, 0, 1, 0 ] ),
    'shape': shape,
    'strides': sm,
    'offset': om,
    'order': 'row-major'
};
var out = {
    'dtype': 'float64',
    'data': new Float64Array( 3 ),
    'shape': [ 3 ],
    'strides': sy,
    'offset': 0,
    'order': 'row-major'
};

mskfilter( [ x, mask, out ] );

console.log( out.data );
// => <Float64Array>[ 2.0, 4.0, 6.0 ]
```

The function accepts the following arguments:

> -   **arrays**: array-like object containing input ndarray, mask ndarray and one dimensional output ndarray.

Reviewer: ```suggestion
-   **arrays**: array-like object containing an input ndarray, a mask ndarray, and a one-dimensional output ndarray.
```

Each provided ndarray should be an object with the following properties:

-   **dtype**: data type.
-   **data**: data buffer.
-   **shape**: dimensions.
-   **strides**: stride lengths.
-   **offset**: index offset.
-   **order**: specifies whether an ndarray is row-major (C-style) or column major (Fortran-style).

</section>

<!-- /.usage -->

<section class="notes">

</section>

<!-- /.notes -->

<section class="examples">

## Examples

```javascript
var discreteUniform = require( '@stdlib/random/base/discrete-uniform' ).factory;
var filledarray = require( '@stdlib/array/filled' );
var filledarrayBy = require( '@stdlib/array/filled-by' );
var ndarray2array = require( '@stdlib/ndarray/base/to-array' );
var mskfilter = require( '@stdlib/ndarray/base/mskfilter' );

var N = 10;
var shape = [ 5, 2 ];
var x = {
    'dtype': 'float64',

>     'data': filledarrayBy( N, 'float64', discreteUniform( -100, 100 ) ),

Reviewer: Let's use random/array/discrete-uniform, random/array/bernoulli, and array/zeros for creating the array buffers.

    'shape': shape,
    'strides': [ 2, 1 ],
    'offset': 0,
    'order': 'row-major'
};
var mask = {
    'dtype': 'uint8',
    'data': filledarrayBy( N, 'uint8', discreteUniform( 0, 1 ) ),
    'shape': shape,
    'strides': [ 2, 1 ],
    'offset': 0,
    'order': 'row-major'
};
var y = {
    'dtype': 'float64',
    'data': filledarray( 0, N, 'float64' ),
    'shape': [ N ],
    'strides': [ 1 ],
    'offset': 0,
    'order': 'row-major'
};

mskfilter( [ x, mask, y ] );
console.log( ndarray2array( x.data, x.shape, x.strides, x.offset, x.order ) );

> console.log( ndarray2array( y.data, y.shape, y.strides, y.offset, y.order ) );

Reviewer: You should also print the mask array to allow users to understand what is being masked.

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# mskfilter

> Apply a mask to a provided input ndarray and assigns to one-dimensional output ndarray.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var mskfilter = require( '@stdlib/ndarray/base/mskfilter' );
```

> #### assign ( arrays )

Reviewer: ```suggestion
#### mskfilter( arrays )
```

> #### mskfilter ( arrays )

Reviewer: ```suggestion
#### mskfilter( arrays )
```

Apply a mask to a provided input ndarray and assign unmasked values to elements in a provided one-dimensional output ndarray.

```javascript
```

The function accepts the following arguments:

-   **arrays**: array-like object containing input ndarray, mask ndarray and one dimensional output ndarray.

Each provided ndarray should be an object with the following properties:

-   **dtype**: data type.
-   **data**: data buffer.
-   **shape**: dimensions.
-   **strides**: stride lengths.
-   **offset**: index offset.
-   **order**: specifies whether an ndarray is row-major (C-style) or column major (Fortran-style).

</section>

<!-- /.usage -->

<section class="notes">

</section>

<!-- /.notes -->

<section class="examples">

## Examples

```javascript
```

</section>

> <!-- /.examples -->

Reviewer: This README is missing the "links" section.
